File: solutions/leetcode/contests/w-436/2-assign_elements_to_groups_with_constraints.scala
object Solution {
    def assignElements(g: Array[Int], e: Array[Int]): Array[Int] = {
        val size = g.length
        val ans = new Array[Int](size)
        
        val mp = e.zipWithIndex.foldLeft(Map[Int, Int]()) { 
            case (m, (v, i)) => if (!m.contains(v)) m + (v -> i) else m 
        }
        
        for (i <- 0 until size) {
            val cur = g(i)
            var min = Int.MaxValue
            
            for (j <- 1 to math.sqrt(cur).toInt) {
                if (cur % j == 0) {
                    if (mp.contains(j)) {
                        min = min.min(mp(j))
                    }
                    val cmp = cur / j
                    if (cmp != j && mp.contains(cmp)) {
                        min = min.min(mp(cmp))
                    }
                }
            }
            
            ans(i) = if (min != Int.MaxValue) min else -1
        }
        ans
    }
}

File: solutions/leetcode/contests/w-436/2-assign_elements_to_groups_with_constraints.java
class Solution {
    public int[] assignElements(int[] groups, int[] elements) {
        int n = groups.length, assigned[] = new int[n];
        Map<Integer, Integer> elementMap = new HashMap<>();
        for (int j = 0; j < elements.length; j++) elementMap.putIfAbsent(elements[j], j);
        
        for (int i = 0; i < n; i++) {
            int currentGroup = groups[i], smallestIndex = Integer.MAX_VALUE;
            for (int j = 1; j * j <= currentGroup; j++) {
                if (currentGroup % j == 0) {
                    if (elementMap.containsKey(j)) smallestIndex = Math.min(smallestIndex, elementMap.get(j));
                    int complement = currentGroup / j;
                    if (elementMap.containsKey(complement)) smallestIndex = Math.min(smallestIndex, elementMap.get(complement));
                }
            }
            assigned[i] = (smallestIndex != Integer.MAX_VALUE) ? smallestIndex : -1;
        }
        return assigned;
    }
}
